### PR TITLE
Added the ec2_run policy

### PR DESCRIPTION
**Pull request by athiruma**
## Type of change
Note: Fill **x** in []
- [ ] bug
- [x] enhancement
- [ ] documentation
- [ ] dependencies

## Description

Added the ec2_run policy.
Created the ADR for the changes

### CODE REVIEW

**Comment by ebattat** [comment]
The best practice is to split this PR to 2:
1. Add ADR
2. Add ec2 run policy
When you are mixing between 2 functions its difficult to debug it in case of errors and also difficult to review it.

File: cloud_governance/common/helpers/aws/aws_cleanup_operations.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    def _delete_resource(self, resource_id: str):
+        """
+        This method deletes the resource by verifying the policy
+        :param resource_id:
+        :type resource_id:
+        :return:
+        :rtype:
+        """
+        action = "deleted"
+        try:
```
Comment (ebattat) [inline cloud_governance/common/helpers/aws/aws_cleanup_operations.py:71]: Should replace to case

File: cloud_governance/common/helpers/aws/aws_cleanup_operations.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        self._iam_client = boto3.client('iam')
+
+    def get_tag_name_from_tags(self, tags: list, tag_name: str) -> str:
+        if tags:
+            for tag in tags:
+                if tag.get('Key').strip().lower() == tag_name.lower():
+                    return tag.get('Value').strip()
+        return ''
+
+    def get_clean_up_days_count(self, tags: list):
```
Comment (ebattat) [inline cloud_governance/common/helpers/aws/aws_cleanup_operations.py:27]: Seems it also updates the cleanup days value ?

File: cloud_governance/common/helpers/aws/aws_cleanup_operations.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        if self._dry_run == 'no':
+            tags = self.__update_tag_value(tags=tags, tag_name='DryRunNoDays', tag_value=str(cleanup_days))
+        else:
+            tags = self.__update_tag_value(tags=tags, tag_name='DryRunYesDays', tag_value=str(cleanup_days))
+        try:
+            if self._policy == 's3_inactive':
+                self._s3_client.put_bucket_tagging(Bucket=resource_id, Tagging={'TagSet': tags})
+            elif self._policy == 'empty_roles':
+                self._iam_client.tag_role(RoleName=resource_id, Tags=tags)
+            elif self._policy in ('ip_unattached', 'unused_nat_gateway', 'zombie_snapshots', 'ebs_unattached', 'ec2_run'):
```
Comment (ebattat) [inline cloud_governance/common/helpers/aws/aws_cleanup_operations.py:140]: What about new policies that we will add in the future line s3_age, do you need add it also here ?
its very risky to tide to specific policy name !!!

File: cloud_governance/main/aws_main_operations.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    def run(self):
+        """
+        This method run the AWS Policy operations
+        :return:
+        :rtype:
+        """
+        policies_list = self.__get_policies()
+        for policy_type, policies in policies_list.items():
+            # @Todo support for all the aws policies, currently supports ec2_run as urgent requirement
+            if self.__policy in policies and self.__policy == "ec2_run":
```
Comment (ebattat) [inline cloud_governance/main/aws_main_operations.py:38]: Why you ask for specific policy  self.__policy == "ec2_run"

File: cloud_governance/main/environment_variables.py [hunk 5]
```diff
@@ -48,7 +52,7 @@ def __init__(self):
             self._environment_variables_dict['account'] = self.get_aws_account_alias_name().upper().replace('OPENSHIFT-', '')
         self._environment_variables_dict['policy'] = EnvironmentVariables.get_env('policy', '')
 
-        self._environment_variables_dict['aws_non_cluster_policies'] = ['ec2_idle', 'ec2_stop', 'ec2_run', 'ebs_in_use',
+        self._environment_variables_dict['aws_non_cluster_policies'] = ['ec2_idle', 'ec2_stop', 'ebs_in_use',
```
Comment (ebattat) [inline cloud_governance/main/environment_variables.py:55]: Why do u remove ec2_run from this list ?

File: cloud_governance/policy/aws/cleanup/ec2_run.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                            'StateTransitionReason': instance.get('StateTransitionReason'),
+                            'RunningDays': running_days,
+                            'CleanUpDays': cleanup_days,
+                            'DryRun': self._dry_run
+                        }
+                    )
+            else:
+                cleanup_days = 0
+                force_tag_update = self._aws_cleanup_policies.get_force_tag_update()
+            self._aws_cleanup_policies.update_resource_tags(resource_id=instance.get('InstanceId'),
```
Comment (ebattat) [inline cloud_governance/policy/aws/cleanup/ec2_run.py:73]: I think the name should be update_resource_day_counts and not all the tags, do you update only the day count ?

File: cloud_governance/policy/policy_operations/aws/zombie_non_cluster/zombie_non_cluster_polices.py [hunk 7]
```diff
@@ -22,32 +22,31 @@ def run(self):
...
-                                logger.info(f'Uploaded the policy results to elasticsearch index: {self._es_index}')
+                if isinstance(response, str):
+                    logger.info(f'key: {cls[0]}, Response: {response}')
+                else:
+                    logger.info(f'key: {cls[0]}, count: {len(response)}, {response}')
+                    policy_result = response
+
+                    if self._es_operations.check_elastic_search_connection():
+                        if policy_result:
+                            if len(policy_result) > 500:
```
Comment (ebattat) [inline cloud_governance/policy/policy_operations/aws/zombie_non_cluster/zombie_non_cluster_polices.py:33]: can you update what the meaning of 500 ?

File: docs/source/index.md [hunk 8]
```diff
@@ -11,7 +11,7 @@ This tool support the following policies:
 
 * Real time Openshift Cluster cost, User cost
 * [ec2_idle](../../cloud_governance/policy/aws/ec2_idle.py): idle ec2 in last 4 days, cpu < 2% & network < 5mb.
-* [ec2_run](../../cloud_governance/policy/aws/ec2_run.py): running ec2.
+* [ec2_run](../../cloud_governance/policy/aws/cleanup/ec2_run.py): running ec2.
```
Comment (ebattat) [inline docs/source/index.md:14]: Why you add cleanup ?

**Comment by athiruma** [comment]
> The best practice is to split this PR to 2:
> 
> 1. Add ADR
> 2. Add ec2 run policy
>    When you are mixing between 2 functions its difficult to debug it in case of errors and also difficult to review it.

I added ADR for the ec2_run policy. It's not the two separate PRs. It was the better time to have the ADR.  Thats why I added the major change.

File: cloud_governance/common/helpers/aws/aws_cleanup_operations.py (hunk 1 above)
Comment (athiruma) [reply cloud_governance/common/helpers/aws/aws_cleanup_operations.py:71]: I wonder, does python have the switch case? 🤔

File: cloud_governance/common/helpers/aws/aws_cleanup_operations.py (hunk 2 above)
Comment (athiruma) [reply cloud_governance/common/helpers/aws/aws_cleanup_operations.py:27]: No, It returns the clean_up days count. We have another method which updates the days count.
[Check here](https://docs.google.com/document/d/1hPv80Xe6-Cao59smi3Yougsa9lfhYparpxVTmNpIM94/edit).

File: cloud_governance/common/helpers/aws/aws_cleanup_operations.py (hunk 3 above)
Comment (athiruma) [reply cloud_governance/common/helpers/aws/aws_cleanup_operations.py:140]: Yes, I know. I was thinking about how I change the process. As of now, we will stick to it later we can replace it.

File: cloud_governance/main/aws_main_operations.py (hunk 4 above)
Comment (athiruma) [reply cloud_governance/main/aws_main_operations.py:38]: As I specified in the new ADR, this is only for ec2_run. So this should work against only ec2_run. I'll activate other policies later. Its risky to activate all policies without testing.

File: cloud_governance/main/environment_variables.py (hunk 5 above)
Comment (athiruma) [reply cloud_governance/main/environment_variables.py:55]: It is the static approach. It is very bad to have like this.
Look this new [ADR Dir structure](https://docs.google.com/document/d/1hPv80Xe6-Cao59smi3Yougsa9lfhYparpxVTmNpIM94/edit). I added **policy/cleanup**, so all cleanup policies will reside in this folder and we will walk through this and run the policies.

File: cloud_governance/policy/aws/cleanup/ec2_run.py (hunk 6 above)
Comment (athiruma) [reply cloud_governance/policy/aws/cleanup/ec2_run.py:73]: Yes, I can rename the method. 👍

File: docs/source/index.md (hunk 8 above)
Comment (athiruma) [reply docs/source/index.md:14]: I moved ec2_run to the cleanup folder. So we have all cleanup policies collectively. [ADR](https://docs.google.com/document/d/1hPv80Xe6-Cao59smi3Yougsa9lfhYparpxVTmNpIM94/edit)

**Pull request merged**